File: algorithms/515. Find Largest Value in Each Tree Row/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def largestValues(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def dfs(root, h):
            if root:
                if h >= len(result):
                    result.append(root.val)
                result[h] = max(result[h], root.val)
                dfs(root.left, h + 1)
                dfs(root.right, h + 1)
        result = []
        dfs(root, 0)
        return result

File: algorithms/515. Find Largest Value in Each Tree Row/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        vector<int> result;
        dfs(root, 0, result);
        return result;
    }

    void dfs(TreeNode * root, int height, vector<int> & result) {
        if ( root ) {
            if ( height < result.size() ) {
                result[height] = max(result[height], root->val);
            } else {
                result.push_back(root->val);
            }
            dfs(root->left, height + 1, result);
            dfs(root->right, height + 1, result);
        }
    }
};
